### PR TITLE
replace shell command with chainable core api

### PR DESCRIPTION
**Pull request by sipsma**
This is one part of our efforts to consolidate dagger CLI commands. It removes `dagger shell` in favor of using a `shell` api chained from the core `Container` type.

For example, previously you'd do:
`dagger shell my-ctr`

Now you do:
`dagger call my-ctr shell`

This required some adjustments in the core API: 
1. `Container` has a new field `withDefaultShell(args: [String!])` that can be used to configure the command to use for opening a shell with the container
1. `shellEndpoint: String!` on `Container` is replaced with `shell(args: [String!]) Terminal!`.
   * If `args` are specified, they override the command to run in the container's shell
   * Otherwise, this defaults to the command specified via `withDefaultShell`
   * As a fallback, `sh` will be executed

---
Implementation details: 

The new `Terminal` type currently only has a field for the existing websocket endpoint, so same basic idea as before in terms of how its implemented. However, because its an object now we have more leeway to easily adjust it in the future if more fields are desired.

This felt like the best solution possible in the short term while keeping options open in the long-term.

`args` is optional and allows specifying the exact command to run in the interactive container. If not set, it first defaults to the entrypoint of the container and falls back to `sh` if those aren't set.

### CODE REVIEW

**Review by helderco — APPROVED** [review]
I love that this process reduces chances of conflict with extra flags in the CLI, moves the intention/verb to the end of the command and generally offloads more into core.

The `TestModuleDaggerShell/args` test failed due to a timeout, but tested the module locally and works well.

**Comment by shykes** [comment]
As you know, I'm fully on board with the idea of using chaining + core API to keep the number of CLI commands and flags as low as possible.

My only note is on the specifics of how `shell` decides what to execute. We already [discussed this in Discord](https://discord.com/channels/707636530424053791/1184618502699614298/1185004131631497267) a while back, resurrecting the relevant parts here:

> shykes:
> Also `Shell` could use a simpler logic for determining what the hell gets executed (something that still confuses me with the current `dagger shell`). I think we should disregard entrypoint altogether, and just make it an argument to `Shell`, with a default value of `sh` if not specified
> Things that would *NOT* affect what `Container.Shell` executes:
> - Past calls to `Container.DefaultArgs`
> - Past calls to `Container.Entrypoint`
> - Past calls to `Container.WithExec`
> - Entrypoint or Command in the base image

> sipsma:
> The one use case I sort of like having some default is if you want to drop right into a repl in the container, in which case it's nice if the container sets up the default for you. But I'm pretty low opinion here; I agree 100% on past calls to Container.WithExec not affecting it, and also agree that DefaultArgs is to confusing. I'd be okay with starting with Entrypoint not having any effect and going from there

> shykes:
> Maybe there’s a `WithDefaultShell`?
> I was brushing up on unix terminology for this stuff
> We could have `Container.Terminal(Shell Optional[string])`
> + `Container.WithDefaultShell`
> `Terminal(): open an interactive terminal to the container`

How do you feel about the above @sipsma ?

**Comment by sipsma** [comment]
> How do you feel about the above @sipsma ?

@shykes Right yeah that's all coming back to me now :-) I like the idea of `Container.WithDefaultShell` and then not letting entrypoint/default-args affect any of this. I'll update with that.

**Comment by sipsma** [comment]
Updated to the `withDefaultShell` approach, PR description has new details on how it works

**Comment by shykes** [comment]
@sipsma how do you feel about this small bikeshedding change:

- Before: `shell: Terminal!`
- After: `terminal: Terminal!`

That clarifies the meaning of "shell" in the API:

- The *terminal* is what you open and connect to for interactive sessions
- The *shell* is the process you execute when opening a terminal

**Comment by sipsma** [comment]
@shykes the overhead of keeping the PR open and rebasing etc. across all the others is getting to be a bit too much so I'm going to merge this one as is w/ `shell`, but happy to continue conversation about switching to `terminal` instead. It will be trivial to do in a followup if we decide to go that way.

File: core/schema/container.graphqls [hunk 1]
```diff
@@ -833,12 +832,18 @@ type Container {
   ): Container!
 
   """
-  Return a websocket endpoint that, if connected to, will start the container with a TTY streamed
-  over the websocket.
-
-  Primarily intended for internal use with the dagger CLI.
+  Return an interactive shell for this container.
   """
-  shellEndpoint: String!
```
[the remark this answers is not in the document]
Comment (sipsma) [reply core/schema/container.graphqls]: Sorry just realized I never got back on this comment, but yes exactly correct.

**Pull request merged**